doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+   list of table names or a special "all tables" publication that gets
+   everything.
+5. **replication slot**: It is an object that keeps track of how far along a
+   replica is with its stream consumption. Normally the upstream database has
+   one such object for each downstream replica. Each replica periodically sends
+   updates with the WAL offset up to which it has consumed which then lets the
+   upstream database safely clean up segments from disk.
+
+### Replication mechanism
+
+In order to replicate a set of tables from an upstream database a replica needs
+to orchestrate two phases, making sure that there is no data loss during the
+handover. The first phase involves the initial table sync which is done with
+normal `COPY` bulk transfer commands. Once the replica has the initial snapshot
+of a table it can switch to reading updates from a replication slot in order to
+keep its table synced.
+
+Ensuring that there is no gap between finishing the initial COPY and starting
+reading from the replication slot is done by synchronizing the database snapshot
+at which the `COPY` is performed with the starting point of the replication
+slot. The simplest way to do this is by using the following query structure:
+
+```sql
+-- Initial sync phase
+BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ
+CREATE_REPLICATION_SLOT <foo> LOGICAL pgoutput USE_SNAPSHOT;
+COPY `table` TO STDOUT;
+COMMIT
+
+-- Streaming phase
+START_REPLICATION SLOT <foo>
+```
+
+By creating the replication slot as the first statement in a transaction and
+setting the `USE_SNAPSHOT` option we ensure that the transaction and the slot
+are at the exact same point. The existence of the slot will prevent any WAL
+segments from being cleaned up while our replica is busy reading the initial
+snapshot.
+
+Once the initial sync is done the transaction is committed to release the
+snapshot and the replica can start streaming the slot normally with no data
+loss.
+
+### Transaction boundaries and timestamps
+
+A key design goal is to maintain the  transaction boundaries from the upstream
+database so that no query on materialize can observe states that never existed
+upstream. At the same time, the data ingested from postgres should be join-able
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
+
+The implementation is greatly simplified by the logical replication protocol
+that guarantees that we'll receive full transactions in commit order. In the
+replication stream each transaction is represented by a BEGIN message, followed
+by one or more row updates, and finally by a COMMIT message.
+
+Each transaction is timestamped with the wall-clock time at the time of the
+BEGIN message. This allows the source implementation to emit rows as it receives
+them and ensure that transaction boundaries are guaranteed.
+
+### Adaptation to multiple tables
```

Comment:
I think you also need to add discussion about how SimpleSource will be extended to support multiple tables?

Reply:
I'm not really sure how this will be done yet and also there is still a possibility of using the `SourceInfo` trait instead if it proves to be easier. I can add this discussion once the implementation has progressed a bit